megamek/src/megamek/common/pathfinder/AeroGroundPathFinder.java

```diff
@@ -0,0 +1,470 @@
+            // if we take more than two, it's a PSR, so definitely avoid that
+            else if(altitudePath.getFinalAltitude() > desiredAltitude &&
+                    altitudePath.getFinalAltitude() >=  startingPath.getFinalAltitude() - 1) {
+                altitudePath.addStep(MoveStepType.DOWN);
+            }
+            // we've either reached the desired altitude or have hit some other stopping condition so that's enough of that
+            else {
+                break;
+            }
+        }
+        
+        return altitudePath;
+    }
+    
+    // the goal here is to generate a "tree-like" structure;
+    //
+    //               \\\|///
+    //                \\|//
+    //                 \|/
+    //                  |
+    //                  |
+    //
+    //  Then, generate smaller subtrees off the left and right sides of the tree, like so:
+    //             /-\\\|///-\
+    //            //--\\|//--\\
+    //                 \|/
+    //                  |
+    //                  |
+    // And so on until we've run out of velocity
+    // if a branch runs off board, we discard it unless it's the shortest one so far, we'll want to keep one of those around
+    // if a branch runs out of velocity, then we make note of all the hexes it has visited so far.
+    // if a hex has already been visited by another branch, then we set the current one as the visitor only if it's shorter
+    private List<MovePath> GenerateAllPaths(MovePath mp) {
+        List<MovePath> retval = new ArrayList<MovePath>();
+        
+        for(MovePath path : generateSidePaths(mp, MoveStepType.TURN_RIGHT)) {
+            // we want to avoid adding paths that don't visit new hexes
+            // off-board paths will get thrown out later
+            if(path.fliesOffBoard() || newHexVisited(path)) {
+                retval.add(path);
+            }
+        }
+        
+        for(MovePath path : generateSidePaths(mp, MoveStepType.TURN_LEFT)) {
+            // we want to avoid adding paths that don't visit new hexes
+            // off-board paths will get thrown out later
+            if(path.fliesOffBoard() || newHexVisited(path)) {
+                retval.add(path);
+            }
+        }
+        
+        ForwardToTheEnd(mp);
+        if(mp.fliesOffBoard() || newHexVisited(mp)) {
+            retval.add(mp);
+        }
+        
+        return retval;
+    }
+    
+    // generates and returns all paths that stick straight lines out of the current path to the right or left
+    private List<MovePath> generateSidePaths(MovePath mp, MoveStepType stepType) {
+        List<MovePath> retval = new ArrayList<MovePath>();
+        MovePath straightLine = mp.clone();
+        
+        while(straightLine.getFinalVelocityLeft() > 0 && 
+                game.getBoard().contains(straightLine.getFinalCoords())) {
+            boolean firstTurn = true;            
```

Comment:
I think you need to declare this outside your loop.

Reply:
That ought to save a couple of cycles generating redundant paths.